cs/src/io/unsafe/OutputPtrBuffer.cs

```diff
@@ -0,0 +1,243 @@
+    /// </summary>
+    public sealed unsafe class OutputPtrBuffer : IOutputStream
+    {
+        internal byte* data;
+        internal int position;
+        internal int length;
+
+        /// <summary>
+        /// Gets data inside the buffer
+        /// </summary>
+        public byte* Data
+        {
+            get { return data; }
+        }
+
+        /// <summary>
+        /// Gets or sets the current position within the buffer
+        /// </summary>
+        public long Position
+        {
+            get { return position; }
+            set { position = checked((int)value); }
+        }
+
+        public OutputPtrBuffer(byte* buffer, int length)
+        {
+            Debug.Assert(BitConverter.IsLittleEndian);
+
+            this.data = buffer;
+            this.length = length;
+
+            position = 0;
+        }
+
+        #region IOutputStream
+
+        /// <summary>
+        /// Write 8-bit unsigned integer
+        /// </summary>
+        public void WriteUInt8(byte value)
+        {
+            if (position >= length)
+            {
+                EndOfStream(sizeof(byte));
+            }
+
+            data[position++] = value;
+        }
+
+        /// <summary>
+        /// Write little-endian encoded 16-bit unsigned integer
+        /// </summary>
+        public void WriteUInt16(ushort value)
+        {
+            if (position + sizeof(ushort) > length)
+            {
+                EndOfStream(sizeof(ushort));
+            }
+
+            var i = position;
+            var b = data;
+            b[i++] = (byte)value;
+            b[i++] = (byte)(value >> 8);
+            position = i;
+        }
+
+        /// <summary>
+        /// Write little-endian encoded 32-bit unsigned integer
+        /// </summary>
+        public void WriteUInt32(uint value)
+        {
+            if (position + sizeof(uint) > length)
+            {
+                EndOfStream(sizeof(uint));
+            }
+
+            var i = position;
+            var b = data;
+            b[i++] = (byte)value;
+            b[i++] = (byte)(value >> 8);
+            b[i++] = (byte)(value >> 16);
+            b[i++] = (byte)(value >> 24);
+            position = i;
+        }
+
+        /// <summary>
+        /// Write little-endian encoded 64-bit unsigned integer
+        /// </summary>
+        public void WriteUInt64(ulong value)
+        {
+            if (position + sizeof(ulong) > length)
+            {
+                EndOfStream(sizeof(ulong));
+            }
+
+            var i = position;
+            var b = data;
+            b[i++] = (byte)value;
+            b[i++] = (byte)(value >> 8);
+            b[i++] = (byte)(value >> 16);
+            b[i++] = (byte)(value >> 24);
+            b[i++] = (byte)(value >> 32);
+            b[i++] = (byte)(value >> 40);
+            b[i++] = (byte)(value >> 48);
+            b[i++] = (byte)(value >> 56);
+            position = i;
```

Comment:
Why not:
```
*((ulong*) (data + position)) = value;
position += sizeof(ulong);
```

The same for other little-endian, fixed-size ints and for `float`/`double`